op/framework.py

```diff
@@ -488,10 +489,23 @@ class SomeObject:
         # TODO Validate that the method has the right signature here.
 
         # TODO Prevent the exact same parameters from being registered more than once.
-
         self._observer[observer.handle.path] = observer
+        weakref.finalize(observer, self._stale_observer_paths.add, observer.handle.path)
         self._observers.append((observer.handle.path, method_name, emitter_path, event_kind))
 
+    def _cleanup_observer_paths(self):
+        if not self._stale_observer_paths:
+            return
+        stale = self._stale_observer_paths
+        new_observers = []
+        for observer_path, method_name, _parent_path, _event_kind in self._observers:
+            if observer_path in stale:
+                continue
+            new_observers.append((observer_path, method_name, _parent_path, _event_kind))
+        # Don't change the set, as the original set.add() is being used in finalizers
+        self._stale_observer_paths.clear()
+        self._observers = new_observers
```

Comment:
What happens if a new observer is added while using the same path as an old observer which is still in the stale list and about to be cleaned. I think this logic would mistakenly de-register the new observer which is alive and awaiting.

Reply:
We could have observe() call self._stale_observer_paths.remove(observer.handle.path) if you feel that is a concern. It depends what concurrency you're worried about. gc could happen at any time, especially if you have reference cycles. So it would be possible to have an object that isn't referenced any more, but gc hasn't triggered. You then call observe replacing that path with a referenced object. And after the new one is registered gc happens.
However:
Do we want to support observe(observer.handle.path) pointing to 2 different objects? My understanding is that while one might be replacing the other, only one of them is ever going to actually get the event, so there really cannot be 2 things registered at the same path.
So we could have something like:
existing = self._observer[observer.handle.path]
if existing is not observer:
  raise RuntimeError()

Thus that sort of race is explicitly considered to be a user error. (If we know about an observer at a path, registering something that is not exactly that object is an error.)

I did try to put together a test that would show off this case. We fundamentally have a bad race condition if you can register an observer at an existing path. Specifically this is the test:
```
# Setup MyNotifier/MyObserver/etc ...

        pub = MyNotifier(framework, "1")
        obs = MyObserver(framework, "2")
        framework.observe(pub.on.foo, obs)
        pub.on.foo.emit()
        self.assertEqual(observed_events, ["foo-2"])
        # Now, we delete the existing obs, which would cause it to be removed from observers,
        # and then immediately register a new obs at the same path using a new object
        del obs
        new_obs = MyObserver(framework, "2")
        framework.observe(pub.on.foo, new_obs)
        pub.on.foo.emit()
        self.assertEqual(observed_events, ["foo-2", "foo-2"])
```

What happens is:
a) My existing code will cause no observed events to occur. Because the newly added observer is flagged as being removed.
b) If I put in a "_stale_observer_paths.remove()" then you end up getting the event *2x* on the new object, because the old _observers and _observer collide. (Because you don't see it as stale, you don't remove the old observer_path, and adding the new observer_path causes it to end up in the list 2x
c) I can just run the 'are there objects to be cleaned up' before we add the new object. This gives us the expected result.

However, there is still a fundamental "when does GC trigger" vs "what step are we in". If we really do want to support multiple registrations at the same path, then probably we want something like this in observe:

```
self._cleanup_stale_observers()
existing = self._observer[path]
if existing is not observer:
  # iterate self._observers and remove all registrations because they point to
  # an object that we are replacing with a new object.
  self._remove_old_observer(path)
self._observer[path] = observer
```

That does have some weird side effects, because you might have registered 2 event handlers for one object, but only one for the second object. But mostly this is a case of 2 objects at the same path is violating the idea of path (IMO).

Thoughts?

Reply:
We can do something like:
```
        for observer_path in self._stale_observer_paths:
            if observer_path in self._observers:
                stale.remove(observer_path)
```

It doesn't avoid ending up having the same path observing 2 times, so pruning before registering the new one should probably be preferred. Another thing that we could do is use something like:
stale.add((path, id(obj))
but then _observers would also need to be
(observer_path, id(observer), ...)
And id(observer) isn't right for how we're doing notifications, deferred into the database, and pulled back out again later. (you *do* want to send the event to the new object after you've recovered from power loss).
I'll mull on it over the weekend.